### ISSUE TITLE
Inaccurate results reported for small methods

### ISSUE THREAD

**Issue by tannergooding** [dotnet/BenchmarkDotNet#1802]
Unless doing instruction level profiling, the highest precision timer on a modern computer is about `25-32 cycles` (which is, under ideal circumstances, about `6ns` on a 5GHz processor at best and `32ns` on a 1GHz processor).

Due to platform specific differences, the maximum reported difference in the high precision timer APIs exposed by the OS is about `100ns`. Additionally, it is well documented that due to the latency between calls and other factors on the OS or hardware, the latency for such a call can be much worse, such as closer to `300ns` when a CPU level timer such as `RDTSC` is not available: https://docs.microsoft.com/en-us/windows/win32/sysinfo/acquiring-high-resolution-time-stamps#resolution-precision-accuracy-and-stability.

While Benchmark.NET does try to account for small methods and while it also tries to account for noise due to call overhead and the like, there are many cases where the numbers it reports are of questionable accuracy.

One such example is the following:
![image](https://user-images.githubusercontent.com/10487869/132735069-fe6e4d5d-f3d4-4f2e-9618-4e3e6147b8eb.png)

In particular, if we look at the first entry `GetShortName_opt` is reporting a time of `0.2082 ns`. Even in an "ideal" scenario where the JIT is able to fully optimize the comparison against a constant value and optimize it to simple be `xor rax, rax`, this is still reporting that it takes approximately 1 cycle on a 5GHz CPU.
* It also shouldn't be able to optimize it like this. AFAIR, Benchmark.NET should be passing the value in and preventing the actual benchmark body from being inlined to avoid such issues.

It would be beneficial, IMO, if Benchmark.NET was more proactive about labeling potentially problematic results and had guidance on how to optimally write a test in a way that will provide accurate results.
* I would view a problematic result, at the very least, as anything taking less than 10ns. Most of these methods should be testing more than a single instruction and are running on 2-4GHz computers. So in an "ideal" environment, 10ns represents no more than 20 instructions and likely no memory accesses. Very few instructions take 0 cycles. Several take 1 cycle and can be pipelined for up to 4 to be in simultaneous dispatch, but its rare to actually have this. Many take 2-3 cycles and if you have any kind of memory access they will take about 3-11 cycles in the fastest scenario (potentially longer for uncached results among other things).

**Comment by timcassell** [comment]
BDN currently under-reports times by 1-2 clock cycles (#1133). Once that is fixed (#2334), I don't think there is any need for more warnings beyond the 0 measurement that we already do.

**Comment by timcassell** [comment]
I'm actually re-opening this because #2336 pretty much solves the issue on x86/x64, but it's still an issue on Arm (at least for Apple M3). https://github.com/dotnet/BenchmarkDotNet/pull/2336#issuecomment-3621826716

**Comment by tannergooding** [comment]
> Interestingly, on Apple M3, an empty method is measured at 0.9ns while a method that increments a field is measured at 0.66ns, which results in the overhead measurement looking "wrong". I don't know if this is specific to M chips or Arm in general, but it doesn't happen on xArch. There does not seem to be a good way for BDN to account for this μarch quirk.

Personally, I would say the fix is to not try and measure method call overhead and instead to encourage users to write benchmarks that take a couple times more than the timer resolution (so more than 100-300ns).

Trying to measure method call overhead is fairly expensive and fundamentally error prone, specifically because we are trying to "guess" how many iterations push us into the realm of "measurable at a level we can then divide to get 'correct' numbers". This guessing gets skewed by things like code or allocation alignment, caching, training of the branch predictor, various other microarchitecture specific optimizations, and far more.

This is the type of issue that we see, and will likely continue to see, every week at scale in the perf triage for the .NET Libraries/Runtime team. Where we regularly get "noisy" results for things from all this inherent skew and the "guess" BDN does for number of iterations to run can get skewed massively by the amount of time that T0 takes to run before it ever gets to T1, among other factors.

And then even with things like "measure 1m empty calls in a loop and then divide by 1m to get the overhead", you're still subject to the cost of the loop itself and other potential aspects, like OS interrupts, hardware delays, other code executing on the machine skewing results, etc. Once you start getting to those finite levels of accuracy, you have to use other hardware specific features to get "correct" measurements, things that BDN largely doesn't have access to (and which typically require admin privileges to query).

**Comment by timcassell** [comment]
> the "guess" BDN does for number of iterations to run can get skewed massively by the amount of time that T0 takes to run before it ever gets to T1

That at least should be greatly improved by #2806 for micro/nano benchmarks.

The rest of your comment I don't disagree, but I defer to @AndreyAkinshin.

**Comment by ladeak** [comment]
I think this issue is a bit misleading and more warnings would be beneficial. For example, I have the following results:

```
// * Summary *

BenchmarkDotNet v0.15.8, Windows 11 (10.0.26200.7309/25H2/2025Update/HudsonValley2)
12th Gen Intel Core i7-1255U 2.60GHz, 1 CPU, 12 logical and 10 physical cores
.NET SDK 10.0.100
  [Host]     : .NET 10.0.0 (10.0.0, 10.0.25.52411), X64 RyuJIT x86-64-v3
  DefaultJob : .NET 10.0.0 (10.0.0, 10.0.25.52411), X64 RyuJIT x86-64-v3


| Method           | Input | Mean      | Error     | StdDev    | Median    | Code Size |
|----------------- |------ |----------:|----------:|----------:|----------:|----------:|
| MultiplyBitShift | 62    | 0.2566 ns | 0.0041 ns | 0.0036 ns | 0.2564 ns |      40 B |
| DivRem           | 62    | 0.0015 ns | 0.0021 ns | 0.0020 ns | 0.0001 ns |      40 B |
| MultiplyBitShift | 100   | 0.2569 ns | 0.0020 ns | 0.0019 ns | 0.2558 ns |      40 B |
| DivRem           | 100   | 0.0000 ns | 0.0000 ns | 0.0000 ns | 0.0000 ns |      40 B |
| MultiplyBitShift | 124   | 0.2094 ns | 0.0025 ns | 0.0022 ns | 0.2089 ns |      40 B |
| DivRem           | 124   | 0.0016 ns | 0.0026 ns | 0.0022 ns | 0.0001 ns |      40 B |

// * Warnings *
ZeroMeasurement
  Divison31.DivRem: Default -> The method duration is indistinguishable from the empty method duration
  Divison31.DivRem: Default -> The method duration is indistinguishable from the empty method duration
  Divison31.DivRem: Default -> The method duration is indistinguishable from the empty method duration
```

While `DisassemblyDiagnoser` reveals that the JITted code is equivalent in both cases:

```
       mov       ecx,[rcx+8]
       mov       edx,84210843
       mov       eax,edx
       imul      ecx
       add       edx,ecx
       mov       eax,edx
       shr       eax,1F
       sar       edx,4
       add       eax,edx
       mov       edx,eax
       shl       edx,5
       sub       edx,eax
       sub       ecx,edx
       sete      al
       movzx     eax,al
       ret
```

EDIT: my confusion is that not even the error or std dev reveals the issue.

**Comment by timcassell** [comment]
@ladeak If the disassembly is truly the same, then the results should be identical. That looks like a different, unrelated issue. Can you try running it on the latest [nightly](https://benchmarkdotnet.org/articles/guides/nuget.html#nightly) version, and open a new issue if it still repros?

**Comment by tannergooding** [comment]
> If the disassembly is truly the same, then the results should be identic

@timcassell. This is the same general issue being reported by me above. BDN has issues measuring small methods and anything within the "general noise" levels of the system. It is not only a fundamental issue with trying to measure something that takes less time than the precision of your timer, but also something that is error prone due to natural fluctuations and general global shared state the system has. It means that the same identical code, with the same alignment, etc can still be measured quite differently at these low scales.

It's very similar to the problem of "measuring the speed of light". Nothing can go faster than the speed of light, so how do you accurately measure it given the timer must be slower than the thing you're measuring? Well, you have to jump through some hoops and get something that is essentially an estimate. A common example is sending a pulse of light to some type of reflector that is far enough away to be measured. Then you take the round trip time and divide it in half. However, even this has problems such as inability to know if light travels the same speed in both directions.

There is many minutia that has to be accounted for and the smaller the scale you get, the harder it is to do. There are also some pretty core staples that representatively show BDN is doing it "wrong" here. The measurement above is a simple example where it's reporting an "impossible" result. The `i7-1255U` has a turbo boost of 4.70 GHz, so `1-cycle` takes around `0.21ns`. Nothing can take less than this unit of time, even with features like pipelining. Even under ideal conditions, where everything is in the cache, all possible pipelined instructions are pipelined, register to register moves are 0 cycles and handled in the renamer, etc... then you are still looking at a minimum of 15 cycles in the above code and so it should be taking a minimum of around 3ns, not the reported `0.21ns`.

The reason this is off by a factor of at least 15 (it's likely more, because we're not going to be ideal and not hitting stable 4.7GHz) is because of the general fluctuations of the system and it leading to mismeasurement of the "overhead" and therefore incorrect subtraction from the reported results. The measured data is small enough (and itself fluctuates enough) that it gets lost in the overhead, because we're doing more overhead work than the data we're trying to measure. -- It's like trying to measure a couple grains of sand that are added onto a small pebble.

**Comment by timcassell** [comment]
I'm not disputing those facts, I'm saying the results should be more accurate with the latest changes. We use direct calls instead of delegates now, which eliminates a heap access and indirections, so I would expect benchmarks with the same asm to result in much more similar measurements. Also back when I had an Intel CPU to test, I saw such tiny benchmarks measured (workload - overhead) at 2-3 clock cycles longer. Not quite the 3ns you expect, but at least you shouldn't see 0 anymore for actual work on that particular CPU.

**Comment by ladeak** [comment]
With the nightly the difference is even larger:

```
BenchmarkDotNet v0.16.0-nightly.20251210.357, Windows 11 (10.0.26200.7309/25H2/2025Update/HudsonValley2)
12th Gen Intel Core i7-1255U 2.60GHz, 1 CPU, 12 logical and 10 physical cores
.NET SDK 10.0.100
  [Host]     : .NET 10.0.0 (10.0.0, 10.0.25.52411), X64 RyuJIT x86-64-v3
  DefaultJob : .NET 10.0.0 (10.0.0, 10.0.25.52411), X64 RyuJIT x86-64-v3


| Method           | Input | Mean      | Error     | StdDev    | Code Size |
|----------------- |------ |----------:|----------:|----------:|----------:|
| MultiplyBitShift | 62    | 0.1147 ns | 0.0062 ns | 0.0052 ns |      40 B |
| DivRem           | 62    | 0.4986 ns | 0.0094 ns | 0.0088 ns |      40 B |
| MultiplyBitShift | 100   | 0.0115 ns | 0.0073 ns | 0.0089 ns |      40 B |
| DivRem           | 100   | 0.4956 ns | 0.0099 ns | 0.0087 ns |      40 B |
| MultiplyBitShift | 124   | 0.1270 ns | 0.0099 ns | 0.0088 ns |      40 B |
| DivRem           | 124   | 0.4876 ns | 0.0083 ns | 0.0078 ns |      40 B |
```

**Comment by timcassell** [comment]
Can you share your benchmark code?

**Comment by ladeak** [comment]
Here is the benchmark (EDIT: please ignore the meaning of the method names in the benchmark).

```csharp
using BenchmarkDotNet.Attributes;
using BenchmarkDotNet.Running;

BenchmarkSwitcher.FromAssembly(typeof(Program).Assembly).Run(args);

[DisassemblyDiagnoser]
public class Divison31
{
    [Params(62, 100, 124)]
    public int Input { get; set; }

    [Benchmark]
    public bool MultiplyBitShift() => (Input / 31) * 31 == Input;

    [Benchmark]
    public bool DivRem() => Input % 31 == 0;
}
```

Here is the assembly code generated on my machine (interestingly, 1 instruction is different this time when I run it (cmp vs sub), but I don't think this should have any meaningful impact as seen above). I am very sure the JITted code was fully identical in my earlier executions - I compared them multiple times.


## .NET 10.0.0 (10.0.0, 10.0.25.52411), X64 RyuJIT x86-64-v3 (Job: DefaultJob)

```assembly
; Divison31.MultiplyBitShift()
       mov       ecx,[rcx+8]
       mov       edx,84210843
       mov       eax,edx
       imul      ecx
       add       edx,ecx
       mov       eax,edx
       shr       eax,1F
       sar       edx,4
       add       eax,edx
       mov       edx,eax
       shl       edx,5
       sub       edx,eax
       cmp       edx,ecx
       sete      al
       movzx     eax,al
       ret
; Total bytes of code 40
```

## .NET 10.0.0 (10.0.0, 10.0.25.52411), X64 RyuJIT x86-64-v3 (Job: DefaultJob)

```assembly
; Divison31.DivRem()
       mov       ecx,[rcx+8]
       mov       edx,84210843
       mov       eax,edx
       imul      ecx
       add       edx,ecx
       mov       eax,edx
       shr       eax,1F
       sar       edx,4
       add       eax,edx
       mov       edx,eax
       shl       edx,5
       sub       edx,eax
       sub       ecx,edx
       sete      al
       movzx     eax,al
       ret
; Total bytes of code 40
```

**Comment by AndreyAkinshin** [comment]
I've been thinking about whether overhead subtraction is still the right approach for BenchmarkDotNet, and I've come to the conclusion that it creates more problems than it solves. I'd like to propose removing it.

**The fundamental problem**

Overhead subtraction works by measuring the per-iteration cost of calling an empty method in a loop and subtracting it from the workload measurement. For this to improve accuracy, the overhead measurement must be more precise than the bias it removes. At the nanosecond scale, this condition does not hold — the measurement error of the overhead is comparable to or larger than the overhead itself. We end up subtracting noise from noise, which amplifies variance rather than reducing it.

This manifests in several observable ways:

- Reported times below one clock cycle, which are physically impossible regardless of pipelining or microarchitectural optimization
- Methods with functionally equivalent machine code producing substantially different results (e.g., see @ladeak's example above where the same computation via `cmp` vs `sub` shows a 171x difference on stable and reverses ranking on nightly)
- On some architectures, the empty method measuring slower than a method that does actual work, producing negative corrections (e.g., the Apple M3 case discussed above)

These aren't quirks of specific benchmarks — they follow directly from trying to measure a signal smaller than the noise floor of the instrument.

**The cost-benefit across measurement ranges**

For **fast methods (< ~10 ns)**: overhead subtraction is actively harmful. The overhead measurement error dominates the result, producing physically impossible numbers, zero measurements, and unstable rankings between runs.

For **medium methods (10–100 ns)**: the correction is a few percent and roughly directionally correct, but the practical benefit is marginal given other sources of variance in real-world measurements (OS scheduling, thermal throttling, turbo state, background processes).

For **slow methods (> 1 μs)**: the correction is negligible — well within noise.

The marginal improvement in the medium range does not justify the catastrophic failures for fast methods and the added engine complexity.

**What changes if we remove it**

All results shift up by a small constant bias — typically 1–3 ns depending on architecture. This means:

- Results are always positive and physically plausible
- The bias is approximately constant across methods, so relative comparisons remain valid
- Consistent results for equivalent code across runs and BDN versions
- Significant simplification of the engine

The tradeoff is that absolute numbers include call overhead. For methods > 1 μs, this is invisible. For fast methods, the numbers are slightly inflated but honest — "4 ns with known constant call overhead" is more useful than "0.2 ns that violates physics."

One consequence is that BDN can no longer distinguish an empty benchmark method from a very fast one — both will report in the 1–3 ns range. This is inherent to removing overhead subtraction, and we should communicate it clearly through a new warning (see plan below).

This is a **breaking change** in reported absolute numbers and should be clearly communicated in release notes. However, relative comparisons — the primary use case — are unaffected.

**This builds on recent progress**

The move to direct calls (#2336) already reduced call overhead significantly. This means the constant bias from *not* subtracting is smaller than ever, while the overhead measurement itself is harder to get right (smaller signal, same noise floor). The less overhead there is, the less value there is in trying to subtract it. Removing overhead subtraction is a natural next step.

**On the useful range of the tool**

I don't think we need to restrict BDN to methods > 100 ns. Without overhead subtraction, the iteration-based approach (run N iterations, divide total time) should produce reliable results well into the single-digit nanosecond range. The sub-nanosecond zone is where precision genuinely breaks down, not the sub-100 ns zone. I'd rather fix the measurement approach than narrow the tool's useful range.

**Proposed plan**

The change can be split into two phases: the core behavioral change, and a follow-up cleanup of now-dead code.

**Phase 1 — Remove overhead subtraction from results (core change):**

1. **`RunResults.GetWorkloadResultMeasurements()`** (`RunResults.cs`): Stop subtracting overhead from workload measurements. Currently this computes the median of overhead measurements and subtracts it from each workload measurement, clamping negative values to zero via `Math.Max(0, ...)`. The `IsSuspiciouslySmall` check (which forces values below 0.1 ns to zero) also becomes unnecessary. The method should pass workload measurements through directly.

2. **`EngineStage.EnumerateStages()`** (`EngineStage.cs`): Skip overhead warmup and overhead actual stages. Currently gated by the `evaluateOverhead` flag (lines 76–80). The `evaluateOverhead` logic in `EnginePilotStageInitial` (lines 68, 81) also becomes unnecessary.

3. **Skip overhead iterations in JIT stages** (`EngineJitStage.cs`): The `EngineFirstJitStage` and `EngineSecondJitStage` currently run overhead iterations alongside workload iterations to ensure the overhead method is JIT-compiled. Without overhead measurement, these can be skipped.

4. **Replace `ZeroMeasurementAnalyser` with `SubNanosecondAnalyser`**: The current analyser (`ZeroMeasurementAnalyser.cs`) compares workload measurements against overhead using a Mann-Whitney equivalence test, and warns "The method duration is indistinguishable from the empty method duration." Without overhead measurements, this approach no longer applies. The replacement should use a simple threshold check on per-operation results. If the result is below ~1 ns, emit a warning explaining that: (a) results at this scale include a constant call overhead bias, (b) the benchmark method body may be empty or doing negligible work, and (c) the user should investigate with `DisassemblyDiagnoser` if they need to determine whether the method body contributes meaningful work. Update the registration in `DefaultConfig.cs` accordingly.

5. **Update the launch count heuristic** (`BenchmarkRunnerClean.cs`): The `AnalyzeLaunchVariance` logic (lines 551–557) currently uses the overhead-to-workload ratio to decide how many launches are needed. Without overhead measurements, this heuristic needs an alternative approach or can be simplified.

**Phase 2 — Deprecate and clean up overhead infrastructure (follow-up):**

6. **Deprecate public API surface**: Mark `EvaluateOverheadAttribute`, `AccuracyMode.EvaluateOverhead`, `Job.WithEvaluateOverhead()`, the `--noOverheadEvaluation` CLI option (`CommandLineOptions.cs`), `IterationMode.Overhead`, and `MeasurementExtensions.IsOverhead()` as `[Obsolete]`. These become no-ops for backward compatibility. Update XML doc comments on `RunStrategy.ColdStart` and `RunStrategy.Monitoring` that reference "overhead evaluating."

7. **Clean up code generation**: Remove `__Overhead`, `OverheadActionUnroll`, and `OverheadActionNoUnroll` from the `BenchmarkType.txt` template, `RunnableEmitter` (IL emit toolchain), `BenchmarkActionFactory`/`BenchmarkActionFactory_Base` (NoEmit toolchain), `RunnableConstants`, and `RunnableReuse`. Remove `OverheadActionNoUnroll`/`OverheadActionUnroll` from `EngineParameters` and the corresponding null checks in the `Engine` constructor. Update `InProcessNoEmitRunner` which creates overhead actions.

8. **Clean up ETW events**: The `EngineEventSource` and its Windows-side parser `EngineEventLogParser` have 6 overhead-specific events (`OverheadJittingStart/Stop`, `OverheadWarmupStart/Stop`, `OverheadActualStart/Stop`, event IDs 3–4, 9–12). These are public API for ETW consumers. Mark as deprecated; keep the event IDs reserved to avoid breaking existing trace parsers.

9. **Update documentation**: The "How it works" guide (`how-it-works.md`) describes the `Result = ActualWorkload - <MedianOverhead>` formula and lists `OverheadWarmup`/`OverheadWorkload` as iteration types. The jobs documentation (`jobs.md`) describes the `EvaluateOverhead` setting and lists `[EvaluateOverhead]` among mutator attributes. Both need to be updated to reflect the new behavior.

**Note on hardware counter diagnostics:**

`TraceLogParser.cs` (Windows PMC diagnostics) uses overhead iterations for a separate purpose — it counts hardware counter samples during overhead windows and subtracts them from workload counters. This is the same overhead subtraction approach applied to PMC data rather than timing. Removing overhead stages will affect this diagnostic. This should be addressed separately, as PMC overhead subtraction is a distinct concern with its own tradeoffs.

I'd like to hear your thoughts. @timcassell @tannergooding

**Comment by timcassell** [comment]
I was planning on utilizing the overhead measurement to deduct the cost of #1782. What if we keep the engine the same, but change the overhead implementation depending on if any overhead besides method invocation needs to be measured? We don't currently have any differing overheads besides the method invocation, so initially it would just be what you say (removing the overhead). Besides #1782, we could also repurpose it for user-provided overhead (#654); users could then optionally provide an empty overhead method to get back the old behavior if they want. We could then also simply remove the clamp to 0, so if the user's overhead is larger than their workload, they get a negative number so it's super obvious that something is wrong.

As for the overhead implementation change, #2958 changes the delegates shape so that the overhead could simply return `ClockSpan.Zero` without actually running anything.

Another thought I have is that the overhead measurement also includes the iteration loop, not just the empty method invocation. So if we remove it, then <10 ns benchmarks will be skewed even greater. Unfortunately I don't think we can simply measure the loop by itself because the JIT could optimize the entire loop away (we can't use `NoOptimization` either because it would not exactly match the workload). I do have an idea to mitigate the issue, though:

```cs
// Pseudo
public ClockSpan OverheadActionNoUnroll(long invokeCount, IClock clock)
{
    long deadCodePreventer = 0;
    return OverheadActionNoUnrollCore(invokeCount, clock, ref deadCodePreventer);
}

[MethodImpl(MethodImplOptions.AggressiveOptimization | MethodImplOptions.NoInlining)]
private ClockSpan OverheadActionNoUnrollCore(long invokeCount, IClock clock, ref long deadCodePreventer)
{
    var startedClock = clock.Start();
    for (long i = 0; i < invokeCount; ++i)
    {
        deadCodePreventer = i;
        // No longer call empty method, we only measure the iteration loop.
        // If the user supplied a [Overhead] method, call that.
    }
    return startedClock.Elapsed;
}

public ClockSpan WorkloadActionNoUnroll(long invokeCount, IClock clock)
{
    long deadCodePreventer = 0;
    return WorkloadActionNoUnrollCore(invokeCount, clock, ref deadCodePreventer);
}

[MethodImpl(MethodImplOptions.AggressiveOptimization | MethodImplOptions.NoInlining)]
private ClockSpan WorkloadActionNoUnrollCore(long invokeCount, IClock clock, ref long deadCodePreventer)
{
    var startedClock = clock.Start();
    for (long i = 0; i < invokeCount; ++i)
    {
        deadCodePreventer = i;
        UserBenchmark();
    }
    return startedClock.Elapsed;
}
```

**Comment by AndreyAkinshin** [comment]
@timcassell Your point about loop overhead is valid — removing overhead entirely loses the ability to correct for the iteration loop cost. But showing negative numbers to users is a UX problem: a negative duration is physically meaningless, and the likely reaction is confusion rather than insight.

I'd like to propose a different approach: **report the result as an interval when measurement uncertainty is high.**

Instead of computing `workload - overhead` as a single number (which can go negative or land below one clock cycle), we compute robust bounds on the shift between the overhead and workload distributions. The result becomes `[0.0; 2.8] ns` instead of `0.2 ns` or `-0.3 ns`.

Concretely:

1. **Keep the overhead measurement** — we still run overhead iterations and collect the data.
2. **Use `ShiftBounds` to compute an interval.** `ShiftBounds` from [Pragmastat](https://pragmastat.dev/) computes distribution-free confidence bounds on the Hodges–Lehmann shift between two samples. Takes the overhead sample, the workload sample, and a `misrate` parameter controlling coverage. Returns `[lower, upper]` — the range of plausible shift values. No distributional assumptions, robust to the exact kind of noise we see here.
3. **Clamp the lower bound to zero.** Negative lower bound means "can't distinguish from zero" — we display `0`, never a negative number.

What this gives us:

- **Honest results.** Instead of `0.2 ns` for code that takes at least ~3 ns, we'd report `[0.0; 3.1] ns`. No physically impossible numbers, no false precision.
- **Relative comparisons still work.** `[0.0; 2.8] ns` vs `[5.4; 8.7] ns` — the difference is clear. Two overlapping intervals near zero — the user correctly sees "we can't distinguish these" without needing to interpret a `ZeroMeasurement` warning.
- **Preserves the overhead infrastructure.** We still collect overhead data, just use it differently. Keeps the door open for #1782 and #654.

For the loop overhead concern: we could combine this with your `deadCodePreventer` approach — measure just the loop (not an empty method call), then feed those measurements into `ShiftBounds` against the workload measurements. Robust loop cost correction without the instability of point-estimate subtraction.

**Comment by tannergooding** [comment]
I'm still in favor of removing it.

Trying to measure the overhead of both the empty method call and "empty loop" are problematic for other reasons as well, such as compiler optimizations, codegen differences, and even differences in how hardware treats such codegen for empty vs non-empty methods, loops, etc. So even if you measure it, it isn't strictly an accurate reading of such overhead and introduces its own problems and complexities.

There really isn't a good way to do such measurements short of running in admin mode and directly querying the model specific registers that allow instruction level profiling and other metadata to be tracked, but all the tools that do that tend to be highly specialized and from the hardware vendors themselves, often with specialized drivers or other software and potentially requiring BIOS/UEFI configuration.

**Comment by tannergooding** [comment]
Just talking about Zen5 (and older AMD CPUs have many similar features as do Intel and Arm64 CPUs), you have these types of branch optimizations:
• next-address logic
• branch target buffer
• return address stack (RAS)
• indirect target predictor
• advanced conditional branch direction predictor
• fetch window tracking structure

These all work together in various interesting ways and make the following two (pseudo-code) samples substantially different to the CPU:
```
loop:
    call empty_method
    cmp counter, limit
    jcc loop

empty_method:
    ret

; vs

loop:
    call actual_method
    cmp counter, limit
    jcc loop

actual_method:
    ; whatever work is being measured, which often has a branch or two, maybe a loop, etc
    ret
```

**Comment by timcassell** [comment]
I have a Zen 5 CPU. I ran the IntroSmokeArguments with and without overhead:

| Method      | Job        | EvaluateOverhead | Mean      | Error     | StdDev    |
|------------ |----------- |----------------- |----------:|----------:|----------:|
| Increment01 | DefaultJob | Default          | 0.0326 ns | 0.0080 ns | 0.0075 ns |
| Increment02 | DefaultJob | Default          | 0.7626 ns | 0.0015 ns | 0.0013 ns |
| Increment03 | DefaultJob | Default          | 0.9565 ns | 0.0038 ns | 0.0036 ns |
| Increment04 | DefaultJob | Default          | 1.1710 ns | 0.0031 ns | 0.0029 ns |
| Increment05 | DefaultJob | Default          | 1.3600 ns | 0.0009 ns | 0.0008 ns |
| Increment06 | DefaultJob | Default          | 1.5497 ns | 0.0013 ns | 0.0012 ns |
| Increment01 | Job-CNOFNS | False            | 0.9925 ns | 0.0190 ns | 0.0211 ns |
| Increment02 | Job-CNOFNS | False            | 1.7179 ns | 0.0024 ns | 0.0023 ns |
| Increment03 | Job-CNOFNS | False            | 1.9117 ns | 0.0030 ns | 0.0026 ns |
| Increment04 | Job-CNOFNS | False            | 2.1245 ns | 0.0015 ns | 0.0013 ns |
| Increment05 | Job-CNOFNS | False            | 2.3134 ns | 0.0022 ns | 0.0017 ns |
| Increment06 | Job-CNOFNS | False            | 2.5037 ns | 0.0036 ns | 0.0032 ns |

Considering each additional increment adds 0.2ns (inc1 being the outlier), I would think the results with overhead subtracted are more accurate.

I also experimented with measuring only the loop, and it resulted in `~0.0120 ns/op`, which doesn't seem worth it at all (the current overhead is `~0.9 ns/op`).

**Comment by tannergooding** [comment]
> I ran the IntroSmokeArguments with and without overhead:

I assume you mean `IntroSmokeIncrements` <https://github.com/dotnet/BenchmarkDotNet/blob/master/samples/BenchmarkDotNet.Samples/IntroSmokeIncrements.cs>?

> Considering each additional increment adds 0.2ns (inc1 being the outlier)

This is not correct. While a standalone `inc reg` takes `1 cycle latency` with a `0.2 cycle reciprocal throughput`, the up to 5x pipelining can only kick in for some non-chained operations and the cycle time is going to be heavily variable based on overall thermal load of the CPU (i.e. you fluctuate between the base clock rate and potential maximum boost clock rate).

But the more important consideration is that we're not really measuring `inc reg` at all because a memory address is involved. That is, only `Increment01` actually does `inc dword ptr [rcx+8]`, while `Increment02+` does `mov eax, dword ptr [rcx+8]` (initial load) followed by  a chain of `inc eax; mov dword ptr [rcx+8], eax` operations. This means that Increment01 is doing fundamentally different work as its an atomic read+write (due to the x64 memory model) while Increment02+ are doing `local mutation, store` chains and not repeatedly loading the value, which changes the potential hardware optimizations that can kick in.

When you factor in the documented latencies, the Zen5 optimization guide details on expected cache latency amounts, etc. You end up getting something that reasonably resembles the above no overhead numbers and such numbers are then mirrored by instruction level profilers like AMD uProf (or Intel VTune for Intel based processors).

-----------------------

On my Zen4 processor (7950X), which has very similar hardware features and instruction cycle timings to Zen5, AMD uProf notes that we take around `1.1 CPI` (Cycles Per Instruction) for `Increment01` and `0.53-0.76 CPI` for `Increment02+`. -- Noting this is utilizing the correct microarchitecture specific features to isolate timings to only the methods in question, so we have no worry about accidental overhead being included. -- Noting that the `1.1 CPI` vs `0.96 CPI` discrepancy in the table is the difference of "just inc" vs also including the "ret"

These numbers then fluctuate a bit based on how the general system is being used. I've shared some basic output from them below. If you dig into the extended info then you can see more info about cache misses, branch mispredicts, and other considerations for all of these.

## Increment01
|Address|Assembly|CYCLES_NOT_IN_HALT|RETIRED_INST|IPC|CPI|
|--|--|--|--|--|--|
|0x0|inc [rcx+0x08]|2,910,250,000|2,581,500,000|1.0353|0.9659|

## Increment02
|Address|Assembly|CYCLES_NOT_IN_HALT|RETIRED_INST|IPC|CPI|
|--|--|--|--|--|--|
|0x0|mov eax, [rcx+0x08]|3,760,000,000|4,908,000,000|1.3053|0.7661|
|0x3|inc eax|13,750,000|13,500,000|0.9818|1.0185|
|0x5|mov [rcx+0x08], eax|1,250,000|1,750,000|1.4000|0.7143|
|0x8|inc eax|1,000,000|1,250,000|1.2500|0.8000|
|0xa|mov [rcx+0x08], eax|8,250,000|6,500,000|0.7273|1.3750|

## Increment06
|Address|Assembly|CYCLES_NOT_IN_HALT|RETIRED_INST|IPC|CPI|
|--|--|--|--|--|--|
|0x0|mov eax, [rcx+0x08]|1,587,250,000|2,502,500,000|1.5766|0.6343|
|0x3|inc eax|25,000,000|45,250,000|1.8100|0.5525|
|0x5|mov [rcx+0x08], eax|5,750,000|7,750,000|1.3478|0.7419|
|0x8|inc eax|5,250,000|8,500,000|1.6190|0.6176|
|0xa|mov [rcx+0x08], eax|8,000,000|12,000,000|1.5000|0.6667|
|0xd|inc eax|51,000,000|70,750,000|1.3873|0.7208|
|0xf|mov [rcx+0x08], eax|7,250,000|12,000,000|1.6552|0.6042|
|0x12|inc eax|58,500,000|84,500,000|1.4444|0.6923|
|0x14|mov [rcx+0x08], eax|2,262,250,000|4,467,000,000|1.9746|0.5064|
|0x17|inc eax|48,000,000|60,000,000|1.2500|0.8000|
|0x19|mov [rcx+0x08], eax|16,750,000|31,000,000|1.8507|0.5403|
|0x1c|inc eax|129,000,000|184,250,000|1.4283|0.7001|
|0x1e|mov [rcx+0x08], eax|96,750,000|140,750,000|1.3942|0.7173|

**Comment by timcassell** [comment]
Ok, you've convinced me.

I still think we should keep the door open for #1782 and #654 and #147, so the overhead mechanism shouldn't be removed entirely, but we can change the default `EvaluateOverhead` to `false`. We can deprecate `WithEvaluateOverhead`, and remove it entirely when #654 is implemented (the user providing their own `[Overhead]` method will enable it). And we can still go with @AndreyAkinshin's `ShiftBounds` suggestion for when overhead is enabled.

**Comment by AndreyAkinshin** [comment]
When overhead is disabled, we can also consider showing a range of values for sub-nanosecond values to emphasize the level of uncertainty. Since there is no baseline to subtract, we can use `CenterBounds` instead of `ShiftBounds` for this case. Btw, the latest Pragmastat is already referenced in `master`, and both functions are ready to be used.

**Comment by tannergooding** [comment]
Definitely fine with there being some switch to show the measured overhead (a different column maybe even) or something similar, I think that's fully reasonable. Same with showing the range of values or anything of the like for small (say <= 10ns) measurements.

**Comment by ladeak** [comment]
I don't know much about the codebase, but as it is marked as up-for-grab, I would assume this is a relatively low-complexity change. In that case, I am happy to tackle it. I do not plan to submit an AI slop PR.

**Comment by ladeak** [comment]
Let me summarize my understanding.

1, I plan to disable overhead runs (this seems pretty straightforward). I also add a remark in the docs of `WithEvaluateOverhead` that it is deprecated (as I notice, I cannot add `[Obsolete]` attribute on it, because it triggers build errors elsewhere). If there is another attribute please let me know. I will not remove any overhead related code as of now (until the above mentioned issues are implemented)

2, With workload measurements only, I will add `CenterBounds` to statistics. What value shall I use for `misrate` (if I should use any)?
This can be displayed as:

<img width="951" height="120" alt="Image" src="https://github.com/user-attachments/assets/62333514-7994-4673-87cc-15c7951f475f" />

(as I understand both ends are inclusive).

3, Center Upper/Lower Bounds can be added to the summary table. In statistics the `StatisticColumn` seems to require a `Func` with a `double` return value, so I am thinking to use 2 columns like this (but only showing them if the `Mean` value is less than 10ns):

<img width="710" height="85" alt="Image" src="https://github.com/user-attachments/assets/401338cf-676c-4dc7-99fb-800714041328" />

Or shall I create a new `IColumn` to display the bounds in a single column? Any preference for the column headers?

4, `RunResult.GetWorkloadResultMeasurements` seems to perform the workload-overhead calculation. I am not certain in what case this is method is actually invoked. Shall I also keep this?

5, As I understand `ShiftBounds` could be used if someone still configures a job with overhead measurements enabled, and both overhead and workload samples are present. `Statistics` types seems to only handle a single simple. So I am not sure this calculation belongs in `Statistics`. I wonder if an analyzer (either changing `ZeroMeasurementAnalyser` or creating a similar one) would be the best way to display the `ShiftBounds` values. What value shall I use for `misrate` parameter in the `ShiftBounds` calculation?

6, Unit tests (mostly integration tests) seems to be a bit slow on my machine. I was searching for some guidelines to find the most relevant ones, but I did not find anything yet. Do you have any suggestions in this area (or shall I rely on the CI for most tests)?

**Comment by timcassell** [comment]
> 1, I plan to disable overhead runs (this seems pretty straightforward). I also add a remark in the docs of `WithEvaluateOverhead` that it is deprecated (as I notice, I cannot add `[Obsolete]` attribute on it, because it triggers build errors elsewhere). If there is another attribute please let me know. I will not remove any overhead related code as of now (until the above mentioned issues are implemented)

`[Obsolete]` is correct. You can suppress the warning for internal code that relies on it.

> 3, Center Upper/Lower Bounds can be added to the summary table. In statistics the `StatisticColumn` seems to require a `Func` with a `double` return value, so I am thinking to use 2 columns like this (but only showing them if the `Mean` value is less than 10ns):
> 
> `Statistics` types seems to only handle a single simple. So I am not sure this calculation belongs in `Statistics`. I wonder if an analyzer (either changing `ZeroMeasurementAnalyser` or creating a similar one) would be the best way to display the `ShiftBounds` values.

It will likely require a larger refactor to support ranges in the summary table. You can just do the overhead changes for now to resolve this issue.

> 4, `RunResult.GetWorkloadResultMeasurements` seems to perform the workload-overhead calculation. I am not certain in what case this is method is actually invoked. Shall I also keep this?

No need to change it. We're only changing the default value of evaluate overhead, all behavior remains the same for now.

> 6, Unit tests (mostly integration tests) seems to be a bit slow on my machine. I was searching for some guidelines to find the most relevant ones, but I did not find anything yet. Do you have any suggestions in this area (or shall I rely on the CI for most tests)?

If you run net462 tests, they are much faster than net8.0 tests. The fastest would be using in-process toolchain in either tfm.

**Comment by ladeak** [comment]
Ok, so not adding the columns to the table nor `CenterBounds` (for now, unless you suggest otherwise).
I have checked the implementation of the `CenterBounds` calculation and realized that value of `misrate` is validated "against" to the sample size, and the sample size must be larger than 1.

**Comment by AndreyAkinshin** [comment]
> It will likely require a larger refactor to support ranges in the summary table. 

I'm doing a large refactoring of summary tables right now. The current implementation will be obsoleted completely, so it doesn't make much sense to deliver improvements to it.
